include/ginkgo/core/base/timer.hpp

```diff
@@ -0,0 +1,185 @@
+Copyright (c) 2017-2023, the Ginkgo authors
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_PUBLIC_CORE_BASE_TIMER_HPP_
+#define GKO_PUBLIC_CORE_BASE_TIMER_HPP_
+
+#include <ginkgo/core/base/executor.hpp>
+
+
+namespace gko {
+
+
+/**
+ * An opaque wrapper for a time point generated by a timer.
+ */
+class time_point {
+public:
+    ~time_point();
+
+    time_point();
+
+    time_point(time_point&&);
+
+    time_point& operator=(time_point&&);
+
+    time_point(const time_point&) = delete;
+
+    time_point& operator=(const time_point&) = delete;
+
+private:
+    friend class Timer;
+    friend class CpuTimer;
+    friend class CudaTimer;
+    friend class HipTimer;
+    friend class DpcppTimer;
+
+    /** What kind of timer was used to generate the time point? */
+    enum class type {
+        /** std::chrono-based timer */
+        cpu,
+        /** cudaEvent-based timer */
+        cuda,
+        /** hipEvent-based timer */
+        hip,
+        /** sycl::event-based timer */
+        dpcpp,
+    };
+
+    type type_;
+    union {
+        CUevent_st* cuda_event;
+        GKO_HIP_EVENT_STRUCT* hip_event;
+        sycl::event* dpcpp_event;
+        std::chrono::steady_clock::time_point chrono;
+    } data_;
+};
+
+
+class Timer {
+public:
+    /**
+     * Creates a new time_point instance for this timer.
+     */
+    virtual void init_time_point(time_point& time) = 0;
```

Comment:
I'll benchmark cudaEventCreate/Destroy to see how much this actually costs us.

Reply:
I would like to avoid allocations in the hot path if possible. The initialization does things like call cudaEventCreate, so they do have a certain overhead

Reply:
I remember now why this was necessary: this is an issue with circular dependencies, because `init_time_point` is implemented in the device backends, but the `time_point` constructors and destructors are implemented in `core`.

Reply:
Thinking about it, the `time_point` constructor should probably not be public, and also just be defaulted. IMO it would be ok to create a time point in the backends, since it is so dependend on it.

Reply:
I have an adequate replacement - we can do the construction in the Timer class and the initializations in the individual backends.